raft-tech/TANF-app
# 3155 prioritized errors

## Summary of Changes
- Added tech memo describing findings on current OFA requirements for prioritizing errors
- Implemented tech memo
  - Updated XLSX generation to use pagination to avoid bringing entire queryset into memory
  - Added queryset filtering based on provided "priority" errors in [this OneNote](https://gorafttech.sharepoint.com/:o:/s/TDRSResearchDesign/EnIa1Mn4v7pOskW7BLomXhIBxUMlYLRU_f1C0dxemW7dWw?e=2ZrCcZ).
  ~~- Updated file status to be partially accepted with errors if more than 500 priority errors are detected~~
  - Updated spreadsheet generator to have a "Critical" and "Summary" tab. Critical is all priority errors and the Summary tab contains all errors grouped by their error message, year, month, and item names/numbers
  - Updated logic so that error report download is user group agnostic

Pull request closes #3155 

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Unzip and submit the attached file `WA_SSPActive_anon.txt`
3. Confirm the status is Partially Accepted with Errors
4. Download the report and verify you get the two tabs discussed above

[WA_SSPActive_anon.txt.zip](https://github.com/user-attachments/files/17464255/WA_SSPActive_anon.txt.zip)

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Tech memo outlining design/method to prioritize errors for reports
+ [x] Memory pressure relieved during report generation no matter the size of the queryset

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)


### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)


### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
